core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java

```diff
@@ -131,13 +137,101 @@ public void heartbeat() throws InterruptedException {
     // Check the set of files that could possibly be over-replicated
     inodes = mInodeTree.getReplicationLimitedFileIds();
     check(inodes, mReplicationHandler, Mode.EVICT);
+
+    // Check the set of files that could possibly be mis-replicated
+    inodes = mInodeTree.getPinIdSet();
+    checkMisreplicated(inodes, mReplicationHandler);
   }
 
   @Override
   public void close() {
     // Nothing to clean up
   }
 
+  /**
+   * Find a set of from and to locations for our moveBlock operation.
+   * After the move, the file will be in the desired pinned medium at least minReplication times.
+   *
+   * @param file the file to scan for misplaced blocks
+   * @param blockInfo blockInfo containing information about the block locations
+   * @return a map that maps from workerHost to desired medium
+   */
+  private Map<String, String> findMisplacedBlock(
+      InodeFile file, BlockInfo blockInfo) {
```

Comment:
Does it make sense to disassociate the `file` concept and simply pass in the pinned medium types and min replication? Also should this deal with max replication or is that handled in another check?

Reply:
this does not handle min replication, there are essentially three checks. One check for minReplication, one check for MaxReplication one check for number of replications that match but medium doesn't match.